cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -453,27 +457,137 @@ class ObjectInputFile final : public io::RandomAccessFile {
 class AzureFileSystem::Impl {
  public:
   io::IOContext io_context_;
-  std::shared_ptr<Azure::Storage::Blobs::BlobServiceClient> service_client_;
+  std::shared_ptr<Azure::Storage::Files::DataLake::DataLakeServiceClient>
+      datalake_service_client_;
+  std::unique_ptr<Azure::Storage::Blobs::BlobServiceClient> blob_service_client_;
   AzureOptions options_;
+  internal::HierarchicalNamespaceDetector hierarchical_namespace_;
```

Comment:
I made is separate because I wanted to keep the cached value `enabled_` private from the rest of `Impl`. I was a bit concerned that people might try to directly access the cached state rather than using the `Enabled()` function. 

I think one possibility is to use a non-smart pointer in `HierarchicalNamespaceDetector` because `HierarchicalNamespaceDetector` will always be destructed at the same time as `Impl`. https://stackoverflow.com/questions/7657718/when-to-use-shared-ptr-and-when-to-use-raw-pointers. I think this would be my preferred solution. What do you think?

Reply:
I decided to just make this change. If you think its a bad idea I'm happy to change it to something else.

Reply:
OK. Let's use the approach.